core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -308,16 +305,17 @@ private[spark] object TaskMetrics extends Logging {
    */
   def fromAccumulators(accums: Seq[AccumulatorV2[_, _]]): TaskMetrics = {
     val tm = new TaskMetrics
-    val (internalAccums, externalAccums) =
-      accums.partition(a => a.name.isDefined && tm.nameToAccums.contains(a.name.get))
-
-    internalAccums.foreach { acc =>
-      val tmAcc = tm.nameToAccums(acc.name.get).asInstanceOf[AccumulatorV2[Any, Any]]
-      tmAcc.metadata = acc.metadata
-      tmAcc.merge(acc.asInstanceOf[AccumulatorV2[Any, Any]])
+    for (acc <- accums) {
+      val name = AccumulatorContext.get(acc.id).flatMap(_.name)
```

Comment:
This code is here 4 times. I am wondering if we should remove name from accumulator metadata and just implement `accumulator.name` using this function.